'if callable(fn)' should infer that 'fn' is not None

Reporting a bug.

Code.
```
fn = None
if callable(fn):
  fn()
```

Actual output.
```
$ mypy  xyz.py 
xyz.py:3: error: "None" not callable
Found 1 error in 1 file (checked 1 source file)
```

Expected behavior.
I would expect `if callable(fn)` to infer a non-None type.

This is a workaround, but shouldn't be necessary:
```
fn = None
if fn and callable(fn):
  fn()
```

mypy version.
```
$ mypy --version
mypy 0.770
$ python --version
Python 3.8.2
```

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -3825,7 +3825,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
         if isinstance(typ, NoneType):
             return [], [typ]
-            
+
         if isinstance(typ, UnionType):
             callables = []
             uncallables = []

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -3823,6 +3823,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         if isinstance(typ, AnyType):
             return [typ], [typ]
 
+        if isinstance(typ, NoneType):
+            return [], [typ]
+            
         if isinstance(typ, UnionType):
             callables = []
             uncallables = []

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -3829,7 +3829,7 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         if isinstance(typ, UnionType):
             callables = []
             uncallables = []
-            for subtype in typ.relevant_items():
+            for subtype in typ.items:
                 # Use unsound_partition when handling unions in order to
                 # allow the expected type discrimination.
                 subcallables, subuncallables = self.partition_by_callable(subtype,

--- a/test-data/unit/check-callable.test
+++ b/test-data/unit/check-callable.test
@@ -449,3 +449,11 @@ if callable():  # E: Too few arguments for "callable"
     pass
 
 [builtins fixtures/callable.pyi]
+
+[case testCallableWithNoneArgs]
+
+fn = None
+if callable(fn):
+  fn()
+
+[builtins fixtures/callable.pyi]

--- a/test-data/unit/check-callable.test
+++ b/test-data/unit/check-callable.test
@@ -457,3 +457,19 @@ if callable(fn):
   fn()
 
 [builtins fixtures/callable.pyi]
+
+[case testCallableUnionOfNoneAndCallable]
+
+from typing import Union, Callable
+
+def f() -> int:
+    return 42
+
+fn = f  # type: Union[None, Callable[[], int]]
+
+if callable(fn):
+    reveal_type(fn) # N: Revealed type is 'def () -> builtins.int'
+else:
+    reveal_type(fn) # N: Revealed type is 'None'
+
+[builtins fixtures/callable.pyi]

